### PR TITLE
Shoelace fixes, adds helper for seeing if you're already interacting with a target

### PR DESCRIPTION
**Pull request by Ryll-Ryll**
## About The Pull Request
You can no longer mess with people's shoelaces from anywhere in the world, and the target moving will properly trigger the hand stomp.

Also adds a list of current do_after targets to /mob and a related helper mob/is_interacting_with(atom/target) which lets you see if you are currently already interacting with a given atom, letting you block stacking if you so choose (or requiring no other interactions in general). This is only used so you can't queue a bunch of shoe unlacing's at once, but feel free to use it elsewhere.

Also adds more stamina damage to getting your hand stomped on by failing shoe-sabotage, and a slightly longer stun.

## Why It's Good For The Game
Fixes and code functionality

## Changelog
:cl: Ryll/Shaps
balance: Failing unlacing someone's shoes and getting your hand stepped on is a bit more punishing
fix: You can no longer untie and knot peoples' shoes from across the world, nor can you queue a bunch of unties to knot them quicker.
fix: Moving while your shoes are being tampered with will properly trigger stamping on the saboteur's hand
/:cl:

### CODE REVIEW

**Comment by trollbreeder** [comment]
Leave the messing with shoelaces anywhere in the world bit, but **ONLY IF YOU HAVE THE TELEKINESIS MUTATION.**

**Comment by Pumpkinoe** [comment]
ADVANCED TK PRANKING

**Comment by Ryll-Ryll** [comment]
I considered that, but you'd be able to cause so much unmitigatable misery by constantly doing it to someone that I figure it's probably for the best to leave it alone (or let someone else do it!)

File: code/__HELPERS/mobs.dm [hunk 1]
```diff
@@ -231,13 +231,19 @@ GLOBAL_LIST_EMPTY(species_list)
...
+	return(target && (target in do_afters))
+
 /proc/do_after(mob/user, var/delay, needhand = 1, atom/target = null, progress = 1, datum/callback/extra_checks = null)
 	if(!user)
 		return 0
 	var/atom/Tloc = null
 	if(target && !isturf(target))
 		Tloc = target.loc
 
+	user.do_afters += target
```
Comment (Rohesie) [inline code/__HELPERS/mobs.dm]:
```suggestion
	LAZYADD(user.do_afters, target)
```

File: code/__HELPERS/mobs.dm [hunk 2]
```diff
@@ -295,6 +301,8 @@ GLOBAL_LIST_EMPTY(species_list)
 				break
 	if (progress)
 		qdel(progbar)
+	if(target)
+		user.do_afters -= target
```
Comment (Rohesie) [inline code/__HELPERS/mobs.dm]:
```suggestion
		LAZYREMOVE(user.do_afters, target)

```

File: code/__HELPERS/mobs.dm [hunk 3]
```diff
@@ -295,6 +301,8 @@ GLOBAL_LIST_EMPTY(species_list)
 				break
 	if (progress)
 		qdel(progbar)
+	if(target)
```
Comment (Rohesie) [inline code/__HELPERS/mobs.dm]:
```suggestion
	if(!QDELETED(target))
```
You'll have to handle removing themselves from this list on `Destroy()`
You can use a signal for that.

File: code/__HELPERS/mobs.dm [hunk 4]
```diff
@@ -231,13 +231,19 @@ GLOBAL_LIST_EMPTY(species_list)
 		checked_health["health"] = health
 	return ..()
 
+///Checks to see if our mob is currently in a do_after with the target as the target
+/mob/proc/is_interacting_with(atom/target)
+	return(target && (target in do_afters))
```
Comment (Rohesie) [inline code/__HELPERS/mobs.dm]:
```suggestion
	return (target in do_afters)
```
If `is_interacting_with()` is invoked with a `target` then there's flawed logic around.

File: code/modules/mob/mob_defines.dm [hunk 5]
```diff
@@ -185,6 +185,9 @@
 	///List of progress bars this mob is currently seeing for actions
 	var/list/progressbars = null	//for stacking do_after bars
 
+	///For storing what do_after's someone has, in case we want to restrict them to only one of a certain do_after at a time
+	var/list/do_afters = list()
```
Comment (Rohesie) [inline code/modules/mob/mob_defines.dm]:
```suggestion
	var/list/do_afters
```
Lazylist.

**Comment by ATH1909** [comment]
> I considered that, but you'd be able to cause so much unmitigatable misery by constantly doing it to someone that I figure it's probably for the best to leave it alone (or let someone else do it!)

Can't they just move to stop your shoelace untying attempt?

File: code/game/atoms.dm [hunk 6]
```diff
@@ -213,6 +216,10 @@
 	LAZYCLEARLIST(overlays)
 	LAZYCLEARLIST(priority_overlays)
 
+	for(var/i in targeted_by)
+		var/mob/M = i
+		LAZYREMOVE(M.do_afters, src)
```
Comment (Rohesie) [inline code/game/atoms.dm]:
```suggestion
	for(var/i in targeted_by)
		var/mob/M = i
		LAZYREMOVE(M.do_afters, src)
	targeted_by = null
```

**Comment by Rohesie** [comment]
Add checks in the `do_after` loop regarding whether the `user` and `target are still referenced by the `do_afters` and `targeted_by` lists.
There's a problem here, however: what if a `user` starts more than one `do_after` action on the same `target`?

File: code/__HELPERS/mobs.dm [hunk 7]
```diff
@@ -231,13 +231,21 @@ GLOBAL_LIST_EMPTY(species_list)
 		checked_health["health"] = health
 	return ..()
 
+///Checks to see if our mob is currently in a do_after with the target as the target
+/mob/proc/is_interacting_with(atom/target)
```
Comment (Rohesie) [inline code/__HELPERS/mobs.dm]: Could you turn this into a macro?
It's not like it's being redefined or anything.
Something like `#define IS_INTERACTING_WITH(source, target)`

File: code/__HELPERS/mobs.dm [hunk 8]
```diff
@@ -231,13 +231,21 @@ GLOBAL_LIST_EMPTY(species_list)
...
 /proc/do_after(mob/user, var/delay, needhand = 1, atom/target = null, progress = 1, datum/callback/extra_checks = null)
 	if(!user)
 		return 0
 	var/atom/Tloc = null
 	if(target && !isturf(target))
 		Tloc = target.loc
 
+	if(target)
+		LAZYADD(user.do_afters, target) // i know technically it already handles if we try to add a null but we're already checking if target exists so shut up
+		LAZYADD(target.targeted_by, user) // i know technically it already handles if we try to add a null but we're already checking if target exists so shut up
```
Comment (Rohesie) [inline code/__HELPERS/mobs.dm]: No need for those comments. We don't want to at null entries to a list, nor we want to runtime. The code is fine.

File: code/__HELPERS/mobs.dm [hunk 9]
```diff
@@ -283,6 +291,10 @@ GLOBAL_LIST_EMPTY(species_list)
 				. = 0
 				break
 
+		if(target && (!(target in user.do_afters) || !(user in target.targeted_by)))
```
Comment (Rohesie) [inline code/__HELPERS/mobs.dm]: No need to check for both. Just check for one of them.
If it happened that one of these checks-if-in-list were true and the other not then there would be a big problem. By default it shouldn't happen.

File: code/modules/clothing/shoes/_shoes.dm [hunk 10]
```diff
@@ -149,53 +149,67 @@
...
+				if(ouchie)
+					ouchie.receive_damage(brute = 10, stamina = 40)
+				L.Paralyze(10)
 
-/**
-  * check_trip runs on each step to see if we fall over as a result of our lace status. Knotted laces are a guaranteed trip, while untied shoes are just a chance to stumble
-  */
+///checking to make sure we're still on the person we're supposed to be, for lacing do_after's
+/obj/item/clothing/shoes/proc/still_shoed(mob/living/carbon/our_guy)
+	return (our_guy == src.loc)
```
Comment (Rohesie) [inline code/modules/clothing/shoes/_shoes.dm]: No need for `src.`

**Review by Rohesie — COMMENTED** [review]
Handle the cleanup of the `do_afters` list on `/mob/Destroy()`

**Comment by Ryll-Ryll** [comment]
bam, should be good @Rohesie

File: code/__HELPERS/mobs.dm (hunk 7 above)
Comment (Rohesie) [reply code/__HELPERS/mobs.dm]: You can delete the proc now that the macro is there.

File: code/modules/clothing/shoes/_shoes.dm [hunk 11]
```diff
@@ -149,53 +149,67 @@
...
+		if(do_after(user, mod_time, needhand=TRUE, target=our_guy, extra_checks=CALLBACK(src, .proc/still_shoed, our_guy)))
 			to_chat(user, "<span class='notice'>You [tied ? "untie" : "knot"] the laces on [loc]'s [src.name].</span>")
 			if(tied == SHOES_UNTIED)
 				adjust_laces(SHOES_KNOTTED, user)
 			else
 				adjust_laces(SHOES_UNTIED, user)
 		else // if one of us moved
-			user.visible_message("<span class='danger'>[our_guy] stamps on [user]'s hand, mid-shoelace [tied ? "knotting" : "untying"]!</span>", "<span class='userdanger'>Ow! [our_guy] stamps on your hand!</span>", user)
-			to_chat(our_guy, "<span class='userdanger'>You stamp on [user]'s hand! What the- they were [tied ? "knotting" : "untying"] your shoelaces!</span>")
+			user.visible_message("<span class='danger'>[our_guy] stamps on [user]'s hand, mid-shoelace [tied ? "knotting" : "untying"]!</span>", "<span class='userdanger'>Ow! [our_guy] stamps on your hand!</span>", list(our_guy))
```
Comment (Rohesie) [inline code/modules/clothing/shoes/_shoes.dm]:
```suggestion
			user.visible_message("<span class='danger'>[our_guy] stomps on [user]'s hand, mid-shoelace [tied ? "knotting" : "untying"]!</span>", "<span class='userdanger'>Ow! [our_guy] stomps on your hand!</span>", list(our_guy))

```

**Pull request merged**